pkg/declextract/interface.go

```diff
@@ -44,3 +47,71 @@ func (ctx *context) finishInterfaces() {
 	}
 	ctx.interfaces = sortAndDedupSlice(ctx.interfaces)
 }
+
+func (ctx *context) processFunctions() {
+	for _, fn := range ctx.Functions {
+		ctx.funcs[fn.File+fn.Name] = fn
+		// Strictly speaking there may be several different static functions in different headers,
+		// but we ignore such possibility for now.
+		if !fn.IsStatic || strings.HasSuffix(fn.File, "*.h") {
+			ctx.funcs[fn.Name] = fn
+		}
+	}
+	nocallers := 0
+	for _, fn := range ctx.Functions {
+		for _, callee := range fn.Calls {
+			called := ctx.findFunc(callee, fn.File)
+			if called == nil || called == fn {
+				continue
+			}
+			fn.calls = append(fn.calls, called)
+			called.callers++
+		}
+		fn.Calls = nil
+		if len(fn.calls) == 0 {
+			nocallers++
+		}
+	}
+}
+
+func (ctx *context) reachableLOC(name, file string) int {
+	fn := ctx.findFunc(name, file)
+	if fn == nil {
+		ctx.warn("can't find function %v in called in %v", name, file)
+		return 0
+	}
+	reachable := make(map[*Function]bool)
+	ctx.collectRachable(fn, reachable)
+	loc := 0
+	for fn := range reachable {
+		loc += fn.LOC
+	}
+	return loc
+}
+
+func (ctx *context) collectRachable(fn *Function, reachable map[*Function]bool) {
+	// Ignore very common functions when computing reachability for complexity analysis.
+	// Counting kmalloc/printk against each caller is not useful (they have ~10K calls).
+	// There are also subsystem common functions (e.g. functions called in some parts of fs/net).
+	// The current threshold is somewhat arbitrary and is based on the number of callers in syzbot kernel:
+	// 6 callers - 2272 functions
+	// 5 callers - 3468 functions
+	// 4 callers - 6295 functions
+	// 3 callers - 16527 functions
+	const commonFuncThreshold = 5
```

Comment:
If you were thinking about kmalloc/printk and related, I'd imagine that a threshold of 10s or 100s would be better suited than just 5 :)

Reply:
There are also subsystem common functions (e.g. functions called in some parts of fs/net).

If we set it to 100, then we count these functions + everything they call 100x. I don't think 100 is appropriate.